python/ray/serve/api.py

```diff
@@ -167,6 +167,14 @@ def _wait_for_goal(self,
 
         ready, _ = ray.wait(
             [self._controller.wait_for_goal.remote(goal_id)], timeout=timeout)
+
+        # Goal as asyncio.Future could return exception if set, ray.get()
+        # retrieves it.
+        try:
+            ray.get(ready)
+        except Exception as e:
+            raise e
+
```

Comment:
ermm this doesn't really do anything does it? it just catches and re-raises.... we can either just pass the exception through or we can get the underlying exception from the object and "prettify" it -- what does the stack trace look like currently?

Reply:
What we currently have here is `ray.wait()`  which doesn't retrieve the exception and fail client goal as expected. It's related to your comment underneath tho about returning exception directly, i can give that a try.